pkg/wallet/account/account.go

```diff
@@ -0,0 +1,265 @@
+package account
+
+import (
+	"crypto/ed25519"
+	"crypto/rand"
+	"fmt"
+
+	"github.com/awnumar/memguard"
+	"github.com/btcsuite/btcutil/base58"
+	"github.com/massalabs/station-massa-wallet/pkg/crypto"
+	"github.com/massalabs/station-massa-wallet/pkg/types"
+	"github.com/massalabs/station-massa-wallet/pkg/types/object"
+	"gopkg.in/yaml.v2"
+)
+
+const (
+	AccountLastVersion = 1
+)
+
+type Account struct {
+	Version      *uint8                    `yaml:"Version"`
+	Nickname     string                    `yaml:"Nickname,omitempty"`
+	Address      types.Address             `yaml:"Address,omitempty"`
+	Salt         [16]byte                  `yaml:"Salt,flow"`
+	Nonce        [12]byte                  `yaml:"Nonce,flow"`
+	CipheredData types.EncryptedPrivateKey `yaml:"CipheredData,flow"`
+	PublicKey    types.PublicKey           `yaml:"PublicKey,flow"`
+}
+
+func New(version *uint8, nickname string, address types.Address, salt [16]byte, nonce [12]byte, cipheredData types.EncryptedPrivateKey, publicKey types.PublicKey) (*Account, error) {
+	if version == nil {
+		version = new(uint8)
+		*version = AccountLastVersion
+	}
+
+	if *version > AccountLastVersion {
+		return nil, fmt.Errorf("unsupported account version: %d", *version)
+	}
+
+	if !NicknameIsValid(nickname) {
+		return nil, fmt.Errorf("invalid nickname: %s", nickname)
+	}
+
+	if err := address.Validate(address.Version, object.UserAddress, object.SmartContractAddress); err != nil {
+		return nil, fmt.Errorf("validating address: %w", err)
+	}
+
+	if err := cipheredData.Validate(cipheredData.Version, object.EncryptedPrivateKey); err != nil {
+		return nil, fmt.Errorf("validating encrypted private key: %w", err)
+	}
+
+	if err := publicKey.Validate(publicKey.Version, object.PublicKey); err != nil {
+		fmt.Println("52", publicKey.Version)
+		fmt.Println("53", publicKey.Kind)
+		return nil, fmt.Errorf("validating public key: %w", err)
+	}
+
+	return &Account{
+		Version:      version,
+		Nickname:     nickname,
+		Address:      address,
+		Salt:         salt,
+		Nonce:        nonce,
+		CipheredData: cipheredData,
+		PublicKey:    publicKey,
+	}, nil
+}
+
+func NewEmpty() Account {
```

Comment:
Why do you need this function?

Reply:
to do this:

```go
acc := account.NewEmpty()
err = acc.Unmarshal(data)
if err != nil {
  return nil, fmt.Errorf("%w: %w", ErrUnmarshalAccount, err)
}
```

Reply:
IMHO this need is the result of the Struct  Account which is haphazard: optional objects and consequential objects (to reduce memory footprint and duplication) should be pointers, whereas objects with default values should be allocations.

Reply:
no need this function

acc := Account{}